packages/taquito-contract-type-generator/src/generator/contract-parser.ts

```diff
@@ -0,0 +1,339 @@
+    }];
+    // throw new GenerateApiError(`Unknown visitVar node: ${JSON.stringify(node, null, 2)} `, { node });
+};
+
+type MType = M.MichelsonType;
+const visitType = (node: MType): TypedType => {
+    // console.log('visitType', { node });
+    // const debug_source = toDebugSource(node);
+
+    // if (typeof node === `string`) {
+    //     return { kind: `value`, raw: node, value: node, typescriptType: `string` };
+    // }
+
+    if (!(`prim` in node)) {
+        // Unknown
+        console.error(`visitType no prim`, { node });
+        return { kind: `unknown`, raw: node };
+    }
+
+    // Union
+    if (node.prim === `or`) {
+        const union = node.args.map(x => visitVar(x)).reduce(reduceFlatMap, []);
+
+        // Flatten
+        const rightSide = union[1];
+        if (rightSide.type.kind === `union`) {
+            union.pop();
+            union.push(...rightSide.type.union);
+        }
+
+        if (union.some(x => !x)) {
+            throw new GenerateApiError(`or: Some fields are null`, { node });
+        }
+        return {
+            kind: `union`,
+            raw: node,
+            union,
+        };
+    }
+
+    // Intersect
+    if (node.prim === `pair`) {
+        const fields = node.args.map(x => visitVar(x)).reduce(reduceFlatMap, []);
+        if (fields.some(x => !x)) {
+            throw new GenerateApiError(`pair: Some fields are null`, { node, args: node.args, fields });
+        }
+        return {
+            kind: `object`,
+            raw: node,
+            fields,
+        };
+    }
+
+    // list
+    if (node.prim === `list`
+        || node.prim === `set`
+    ) {
+        if (node.args.length !== 1) {
+            throw new GenerateApiError(`list does not have 1 arg`, { node, args: node.args });
+        }
+
+        const arrayItem = visitType(node.args[0]);
+        if (!arrayItem) {
+            throw new GenerateApiError(`arrayItem are null`, { node, args: node.args, arrayItem });
+        }
+        return {
+            kind: `array`,
+            raw: node,
+            array: { item: arrayItem },
+        };
+    }
+
+    // map
+    if (node.prim === `map`
+        || node.prim === `big_map`
+    ) {
+        if (node.args.length !== 2) {
+            throw new GenerateApiError(`map does not have 2 args`, { node, args: node.args });
+        }
+
+        const mapKey = visitType(node.args[0]);
+        const mapValue = visitType(node.args[1]);
+        if (!mapKey || !mapValue) {
+            throw new GenerateApiError(`map is missing key or value`, { node, args: node.args, mapKey, mapValue });
+        }
+        return {
+            kind: `map`,
+            raw: node,
+            map: {
+                key: mapKey,
+                value: mapValue,
+                isBigMap: node.prim === `big_map`,
+            },
+        };
+    }
+
+    // option
+    if (node.prim === `option`) {
+        return {
+            ...visitType(node.args[0]),
+            optional: true,
```

Comment:
In Michelson we can have an optional parameter before a required one, but for these cases, when generating the type, typescript shows the following error: 
```
 set_child_record: (
        address?: address,
        data: MMap<string, ( ... // A required parameter cannot follow an optional parameter.
```
Could we return the type or undefined instead of making the parameter optional?
```
 set_child_record: (
        address: address | undefined,
        data: MMap<string, ( ...
```
When I call this method without an address, I explicitly need to write undefined (or null) for this parameter:
```
await c.methods.set_child_record(
          undefined, // address
          new MichelsonMap(), // data
```
so while getting rid of the typescript error, this change could be more intuitive in showing how to write the optional parameter.

Reply:
@roxaneletourneau good catch, I initially had those inside an object and didn't find that when changing it.

Reply:
@roxaneletourneau though I would argue that taquito would work better if the method calls were wrapped in an object (though not backward compatible, but maybe something to consider for future design)